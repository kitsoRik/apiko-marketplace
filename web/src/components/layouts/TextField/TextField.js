import React, { useState, useEffect } from 'react';

import ViewPassword from '../../../assets/icons/view-password.svg';
import ViewPasswordChecked from '../../../assets/icons/view-password-checked.svg';

import ModalLoading from '../ModalLoading/ModalLoading';

import _ from 'lodash';

import './TextField.scss';

const TextField = ({
    className,
    type = "big",
    icon,
    children,
    name,
    loading = false,
    multiline = false,
    autoCompleteOptions,
    autoCompleteOptionsWhenEmptyHeader = null,
    autoCompleteOptionsWhenEmpty = null,
    value = "", password, error, errorIfTouched, onChange = () => { }, onValueChange = () => { }, ...props }) => {
    const [touched, setTouched] = useState(false);
    const [viewPassword, setViewPassword] = useState(false);

    const [autocompleteIndex, setAutocompleteIndex] = useState(-1);

    const [inFocus, setInFocus] = useState(false);

    const err = errorIfTouched && touched;


    useEffect(() => {
        setAutocompleteIndex(-1);
    }, [value])

    const onKeyDown = (e) => {
        switch (e.key) {
            case "ArrowDown": {
                e.preventDefault();
                if (autocompleteIndex === autoCompleteOptions?.length - 1 || autocompleteIndex === autoCompleteOptionsWhenEmpty?.length - 1) {
                    setAutocompleteIndex(0);
                    break;
                }
                setAutocompleteIndex(autocompleteIndex + 1);
                break;
            }
            case "ArrowUp": {
                if (autocompleteIndex === -1) break;
                e.preventDefault();
                setAutocompleteIndex(autocompleteIndex - 1);
                break;
            }
            case "Enter": {
                if (!autoCompleteOptions && !autoCompleteOptionsWhenEmpty) return;
                e.preventDefault();
                let _value;
                if (value === "") {
                    const el = autoCompleteOptionsWhenEmpty[autocompleteIndex];
                    if (!el) return;
                    const props = el.props;
                    _value = props.textValue;
                    if (props.onSelect) props.onSelect();
                }
                else {
                    const el = autoCompleteOptions[autocompleteIndex];
                    if (!el) return;
                    const props = el.props;
                    _value = props.textValue;
                    if (props.onSelect) props.onSelect();
                }
                onValueChange(_value);

                inputRef.current.value = _value;

                setAutocompleteIndex(-1);
                break;
            }
            default: break;
        }
    }

    const inputRef = React.createRef();
    const autoCompleteTruthOptions = (autoCompleteOptions ?? [])
        .filter(({ props: { textValue } }) => new RegExp(value).test(textValue) && value !== textValue);


    let inputComponent = null;

    if (multiline) {
        inputComponent = <textarea
            ref={inputRef}
            style={{ paddingLeft: `${icon ? 40 : 13}px` }}
            className="text-field-"
            error={(err || error) ? "true" : "false"}
            type={password && !viewPassword ? 'password' : 'text'}
            onBlur={() => setTouched(true)}
            value={value}
            name={name}
            onKeyDown={onKeyDown}
            onChange={(e) => { onValueChange(e.target.value); onChange(e); onValueChange(e.target.value); }}
            {...props} >

        </textarea>
    } else {
        inputComponent = <input
            ref={inputRef}
            style={{ paddingLeft: `${icon ? 40 : 13}px` }}
            className="text-field-"
            error={(err || error) ? "true" : "false"}
            type={password && !viewPassword ? 'password' : 'text'}
            onBlur={() => setTouched(true)}
            value={value}
            name={name}
            onKeyDown={onKeyDown}
            onChange={(e) => { onValueChange(e.target.value); onChange(e); onValueChange(e.target.value); }}
            {...props} />
    }

    return (
        <div className={`text-field ${className}`} type={type} onFocus={() => setInFocus(true)} onBlur={_.debounce(() => setInFocus(false), 100)} tabIndex="1">
            {icon &&
                <div className="text-field-icon">
                    {icon}
                </div>
            }
            {inputComponent}
            {password && <img
                className="text-field-view-password"
                alt="View password"
                src={!viewPassword ? ViewPassword : ViewPasswordChecked}
                onClick={() => setViewPassword(!viewPassword)} />}
            {autoCompleteTruthOptions.length !== 0 && inFocus && value !== "" &&
                <div className="text-field-auto-complete-container" tabIndex="1">
                    {
                        (() => {
                            return autoCompleteTruthOptions.map((e, i) => {
                                return React.cloneElement(e, {
                                    active: autocompleteIndex === i,
                                    compareValue: value
                                })
                            });
                        })()
                    }
                </div>
            }
            {autoCompleteOptionsWhenEmpty && autoCompleteOptionsWhenEmpty.length !== 0 && inFocus && value === "" &&
                <div className="text-field-auto-complete-container" tabIndex="1">
                    {autoCompleteOptionsWhenEmptyHeader}
                    {
                        autoCompleteOptionsWhenEmpty.map((e, i) => {
                            return React.cloneElement(e, {
                                active: autocompleteIndex === i
                            })
                        })
                    }
                </div>
            }
            {loading &&
                <div className="text-field-loading">
                    <ModalLoading darken={false} />
                </div>
            }
        </div >
    );
}

export default TextField;