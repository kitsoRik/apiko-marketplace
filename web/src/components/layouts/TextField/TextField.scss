.text-field {
    display: flex;
    align-items: center;

    position: relative;

    &- {
        width: 100%;
        height: 100%;

        flex: 1;

        color: #000;

        font-size: 16px;

        border-radius: 5px;
        border: 1px solid #dedee0;
        box-sizing: border-box;

        resize: none;

        &::placeholder {
            color: #66666666;
        }

        &[error="true"] {
            border-color: red;
        }
    }

    &-icon {
        height: 18px;
        width: 18px;

        position: absolute;

        display: flex;
        align-items: center;
        margin-left: 10px;
    }

    &-loading {
        height: 18px;
        width: 18px;

        position: absolute;
        right: 18px;

        display: flex;
        align-items: center;
        margin-left: 10px;
    }

    &-view-password {
        margin-right: 20px;

        width: 24px;
        height: 16px;

        position: absolute;

        right: 0;
    }

    &-medium {
        &- {
            padding: 9px;
        }
    }

    &-auto-complete-container {
        padding: 15px;
        width: 100%;

        position: absolute;
        top: 100%;

        background-color: white;

        border: 1px solid gray;
        border-radius: 4px;

        box-sizing: border-box;

        z-index: 1500;
    }

    &[type="medium"] .text-field- {
        padding: 9px;
        font-size: 14px;
    }

    &[type="big"] .text-field- {
        padding: 20px 13px;
    }
}
