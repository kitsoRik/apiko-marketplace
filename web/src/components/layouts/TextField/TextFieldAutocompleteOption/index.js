import TextFieldAutocompleteOption from "./TextFieldAutocompleteOption";

export default TextFieldAutocompleteOption;