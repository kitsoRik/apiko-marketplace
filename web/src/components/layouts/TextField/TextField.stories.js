import React from 'react';
import TextField from '.';
import TextFieldAutocompleteOption from './TextFieldAutocompleteOption/TextFieldAutocompleteOption';

export default { title: "TextField" };

export const withPlaceholder = () => <TextField placeholder="My custom placholder" />
export const withValue = () => <TextField value="My custom value" />
export const password = () => <TextField password={true} />
export const withLoading = () => <TextField loading={true} />

export const withAutocmoplete = () => <TextField placeholder="123" autoCompleteOptions={[
    <TextFieldAutocompleteOption value={"Ternopll"} />,
    <TextFieldAutocompleteOption value={"Kyev"} />
]} />