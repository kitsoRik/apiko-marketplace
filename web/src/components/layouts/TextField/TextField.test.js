import React from 'react';
import { create } from 'react-test-renderer';
import TextField from './TextField';

describe("TextField", () => {
    it("should render without text", () => {
        const component = create(<TextField />);
        const json = component.toJSON();

        expect(json).toMatchSnapshot();
    });

    it("should render with icon", () => {
        const component = create(<TextField icon={"1"} />);
        const json = component.toJSON();

        expect(json).toMatchSnapshot();
    });

    it("should render with text", () => {
        const component = create(<TextField value="My custom value" />);
        const json = component.toJSON();

        expect(json).toMatchSnapshot();
    });

    it("should render with text and icon", () => {
        const component = create(<TextField icon={"1"} value="My custom value" />);
        const json = component.toJSON();

        expect(json).toMatchSnapshot();
    });

    it("should render with placeholder", () => {
        const component = create(<TextField icon={"1"} placeholder="My custom placeholder" />);
        const json = component.toJSON();

        expect(json).toMatchSnapshot();
    });

    it("should render with placeholder and icon", () => {
        const component = create(<TextField icon={"1"} placeholder="My custom placeholder" />);
        const json = component.toJSON();

        expect(json).toMatchSnapshot();
    });

    it("should render with placeholder and text", () => {
        const component = create(
            <TextField
                value={"My custom value"}
                placeholder="My custom placeholder" />
        );
        const json = component.toJSON();

        expect(json).toMatchSnapshot();
    });

    it("should render with type=password", () => {
        const component = create(
            <TextField password={true} />
        );
        const json = component.toJSON();

        expect(json).toMatchSnapshot();
    });
});