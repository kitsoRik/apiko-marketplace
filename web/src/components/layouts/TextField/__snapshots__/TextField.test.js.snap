// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField should render with icon 1`] = `
<div
  className="text-field undefined"
  onBlur={[Function]}
  onFocus={[Function]}
  tabIndex="1"
  type="big"
>
  <div
    className="text-field-icon"
  >
    1
  </div>
  <input
    className="text-field-"
    error="false"
    onBlur={[Function]}
    onChange={[Function]}
    onKeyDown={[Function]}
    style={
      Object {
        "paddingLeft": "40px",
      }
    }
    type="text"
    value=""
  />
</div>
`;

exports[`TextField should render with placeholder 1`] = `
<div
  className="text-field undefined"
  onBlur={[Function]}
  onFocus={[Function]}
  tabIndex="1"
  type="big"
>
  <div
    className="text-field-icon"
  >
    1
  </div>
  <input
    className="text-field-"
    error="false"
    onBlur={[Function]}
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="My custom placeholder"
    style={
      Object {
        "paddingLeft": "40px",
      }
    }
    type="text"
    value=""
  />
</div>
`;

exports[`TextField should render with placeholder and icon 1`] = `
<div
  className="text-field undefined"
  onBlur={[Function]}
  onFocus={[Function]}
  tabIndex="1"
  type="big"
>
  <div
    className="text-field-icon"
  >
    1
  </div>
  <input
    className="text-field-"
    error="false"
    onBlur={[Function]}
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="My custom placeholder"
    style={
      Object {
        "paddingLeft": "40px",
      }
    }
    type="text"
    value=""
  />
</div>
`;

exports[`TextField should render with placeholder and text 1`] = `
<div
  className="text-field undefined"
  onBlur={[Function]}
  onFocus={[Function]}
  tabIndex="1"
  type="big"
>
  <input
    className="text-field-"
    error="false"
    onBlur={[Function]}
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="My custom placeholder"
    style={
      Object {
        "paddingLeft": "13px",
      }
    }
    type="text"
    value="My custom value"
  />
</div>
`;

exports[`TextField should render with text 1`] = `
<div
  className="text-field undefined"
  onBlur={[Function]}
  onFocus={[Function]}
  tabIndex="1"
  type="big"
>
  <input
    className="text-field-"
    error="false"
    onBlur={[Function]}
    onChange={[Function]}
    onKeyDown={[Function]}
    style={
      Object {
        "paddingLeft": "13px",
      }
    }
    type="text"
    value="My custom value"
  />
</div>
`;

exports[`TextField should render with text and icon 1`] = `
<div
  className="text-field undefined"
  onBlur={[Function]}
  onFocus={[Function]}
  tabIndex="1"
  type="big"
>
  <div
    className="text-field-icon"
  >
    1
  </div>
  <input
    className="text-field-"
    error="false"
    onBlur={[Function]}
    onChange={[Function]}
    onKeyDown={[Function]}
    style={
      Object {
        "paddingLeft": "40px",
      }
    }
    type="text"
    value="My custom value"
  />
</div>
`;

exports[`TextField should render with type=password 1`] = `
<div
  className="text-field undefined"
  onBlur={[Function]}
  onFocus={[Function]}
  tabIndex="1"
  type="big"
>
  <input
    className="text-field-"
    error="false"
    onBlur={[Function]}
    onChange={[Function]}
    onKeyDown={[Function]}
    style={
      Object {
        "paddingLeft": "13px",
      }
    }
    type="password"
    value=""
  />
  <img
    alt="View password"
    className="text-field-view-password"
    onClick={[Function]}
    src="view-password.svg"
  />
</div>
`;

exports[`TextField should render without text 1`] = `
<div
  className="text-field undefined"
  onBlur={[Function]}
  onFocus={[Function]}
  tabIndex="1"
  type="big"
>
  <input
    className="text-field-"
    error="false"
    onBlur={[Function]}
    onChange={[Function]}
    onKeyDown={[Function]}
    style={
      Object {
        "paddingLeft": "13px",
      }
    }
    type="text"
    value=""
  />
</div>
`;
